static/src/javascripts/projects/common/modules/article/spacefinder.js

```diff
@@ -25,37 +22,91 @@ define([
-    chain,
-    forOwn) {
+    forOwn,
+    curry
+) {
     // find spaces in articles for inserting ads and other inline content
-    var bodySelector = '.js-article__body',
-        // minAbove and minBelow are measured in px from the top of the paragraph element being tested
-        defaultRules = { // these are written for adverts
-            minAbove: 250, // minimum from para to top of article
-            minBelow: 300, // minimum from (top of) para to bottom of article
-            clearContentMeta: 50, // vertical px to clear the content meta element (byline etc) by. 0 to ignore
-            selectors: { // custom rules using selectors. format:
-                //'.selector': {
-                //   minAbove: <min px above para to bottom of els matching selector>,
-                //   minBelow: <min px below (top of) para to top of els matching selector> }
-                ' > h2': {minAbove: 0, minBelow: 250}, // hug h2s
-                ' > *:not(p):not(h2)': {minAbove: 25, minBelow: 250} // require spacing for all other elements
-            }
-        },
-        // test one element vs another for the given rules
-        _testElem = function (para, other, rules) {
-            var isMinAbove = para.top - other.bottom >= rules.minAbove,
-                isMinBelow = other.top - para.top >= rules.minBelow;
-            return isMinAbove || isMinBelow;
+    // minAbove and minBelow are measured in px from the top of the paragraph element being tested
+    var defaultRules = { // these are written for adverts
+        bodySelector: '.js-article__body',
+        slotSelector: ' > p',
+        minAbove: 250, // minimum from para to top of article
+        minBelow: 300, // minimum from (top of) para to bottom of article
+        clearContentMeta: 50, // vertical px to clear the content meta element (byline etc) by. 0 to ignore
+        selectors: { // custom rules using selectors. format:
+            //'.selector': {
+            //   minAbove: <min px above para to bottom of els matching selector>,
+            //   minBelow: <min px below (top of) para to top of els matching selector> }
+            ' > h2': {minAbove: 0, minBelow: 250}, // hug h2s
+            ' > *:not(p):not(h2)': {minAbove: 25, minBelow: 250} // require spacing for all other elements
         },
-        // test one element vs an array of other elements for the given rules
-        _testElems = function (para, others, rules) {
-            return chain(others).and(every, function (other) {
-                return _testElem(para, other, rules);
-            }).valueOf();
-        };
+        // filter:(slot:Element, index:Integer, slots:Collection<Element>) -> Boolean
+        // will run each slot through this fn to check if it must be counted in
+        filter: null,
+        // startAt:Element
+        // will remove slots before this one
+        startAt: null,
+        // stopAt:Element
+        // will remove slots from this one on
+        stopAt: null,
```

Comment:
I think it makes sense if you think backwards: (1) the default case is where `stopAt` is null and all the remaining elements are included and (2) the case where `stopAt` is the last slot and all the remaining elements but this one are included. It almost sounds like a recursive definition :-)